### PR TITLE
repository-service-controller architecture in users

### PR DESCRIPTION
**Pull request by abubakar-sazim**

### CODE REVIEW

File: backend/nest-app/src/auth/auth.service.ts [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
 export class AuthService {
-  signup() {
-    return { msg: 'Signing Up' };
-  }
+  constructor(private usersService: UsersService) {}
 
-  signin() {
-    return { msg: 'Signing In' };
+  async validateUser(email: string, pass: string): Promise<any> {
+    const user = await this.usersService.findOneById(email);
```
Comment (jUddinSazim) [inline backend/nest-app/src/auth/auth.service.ts:11]: `findOneById()` is expecting to have one number arg, but you have pass a string arg, also that is email. May be you can create a new method like `findOneByEmail` or make the `findOne` method to dynamically choose use by either id or email.

File: backend/nest-app/src/auth/auth.service.ts [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
 @Injectable()
 export class AuthService {
-  signup() {
-    return { msg: 'Signing Up' };
-  }
+  constructor(private usersService: UsersService) {}
 
-  signin() {
-    return { msg: 'Signing In' };
+  async validateUser(email: string, pass: string): Promise<any> {
```
Comment (jUddinSazim) [inline backend/nest-app/src/auth/auth.service.ts:10]: it highly encouraged not to use `any` types. Like for you case:
```suggestion
  async validateUser(email: string, pass: string): Promise<Omit<User, 'password'> | null> {
```

File: backend/nest-app/src/users/users.controller.ts
Comment (jUddinSazim) [inline backend/nest-app/src/users/users.controller.ts:1]: resource controllers have standard naming conventions for it's method as well as the API routes.
A helpful article:  https://medium.com/@mehmetbaz/rest-api-naming-conventions-and-best-practices-31635ee11c0d

Method naming convention:
image

File: backend/nest-app/src/users/users.repository.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  async findOneById(id: number): Promise<User | null> {
+    const userRepository = this.orm.em.getRepository(User);
+    const user = await userRepository.findOne({ id });
+    if (!user) {
+      throw new NotFoundException(`User with ID ${id} not found`);
+    }
+    return user;
+  }
+
+  async findUserByEmail(email: string): Promise<User | null> {
```
Comment (jUddinSazim) [inline backend/nest-app/src/users/users.repository.ts:23]:
```suggestion
  async findOneByEmail(email: string): Promise<User | null> {
```

File: backend/nest-app/src/users/users.repository.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    const userRepository = this.orm.em.getRepository(User);
+    const user = await userRepository.findOne({ email });
+    if (!user) {
+      throw new NotFoundException(`User with ID ${email} not found`);
+    }
+    return user;
+  }
+
+  async updateUser(id: number, updateUserDto: UpdateUserDto): Promise<string> {
```
Comment (jUddinSazim) [inline backend/nest-app/src/users/users.repository.ts:32]:
```suggestion
  async update(id: number, data: Partial<User>): Promise<User> {
```
DTO has only relations with the controller, not inside repository layer

File: backend/nest-app/src/users/users.repository.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    const userRepository = this.orm.em.getRepository(User);
+    const user = await userRepository.findOne({ id });
+    if (!user) {
+      return `User with ID ${id} not found`;
+    }
+    userRepository.assign(user, updateUserDto);
+    await userRepository.flush();
+    return `User with ID ${id} updated successfully`;
```
Comment (jUddinSazim) [inline backend/nest-app/src/users/users.repository.ts:40]: after creating, updating and deleting, always return the affected row, i.e. `user` in this case 
```suggestion
    return user;
```

File: backend/nest-app/src/users/users.repository.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { MikroORM } from '@mikro-orm/core';
+import { User } from './users.entity';
+import { UpdateUserDto } from './dto/update-user.dto';
+import { NotFoundException } from '@nestjs/common';
+
+export class UserRepository {
+  constructor(private readonly orm: MikroORM) {}
+
+  async findAllUsers(): Promise<User[]> {
```
Comment (jUddinSazim) [inline backend/nest-app/src/users/users.repository.ts:9]:
```suggestion
  async findAll(): Promise<User[]> {
```

File: backend/nest-app/src/users/users.service.ts [hunk 7]
```diff
@@ -1,33 +1,30 @@
...
+    private readonly userRepository: UserRepository,
+  ) {}
 
   async getAllUsers(): Promise<User[]> {
-    const userRepository = this.orm.em.getRepository(User);
-    return userRepository.findAll();
+    return this.userRepository.findAllUsers();
   }
 
   async findOneById(id: number) {
```
Comment (jUddinSazim) [inline backend/nest-app/src/users/users.service.ts:19]:
```suggestion
  async getUser(id: number) {
```
findOneById etc are conventions for the repository layer. Service layer is more human friendly and business model related.

File: backend/nest-app/src/users/users.service.ts [hunk 8]
```diff
@@ -1,33 +1,30 @@
...
-    const userRepository = this.orm.em.getRepository(User);
-    return userRepository.findOne({ id });
+    return this.userRepository.findOneById(id);
   }
 
-  update(id: number, updateUserDto: UpdateUserDto) {
-    return `This action updates a #${id} user`;
+  async findUserByEmail(email: string) {
```
Comment (jUddinSazim) [inline backend/nest-app/src/users/users.service.ts:23]: this can be implemented dynamically insider `getUser()`

File: backend/nest-app/temp/User.js.json
Comment (jUddinSazim) [inline backend/nest-app/temp/User.js.json:1]: why do we need this temp json file?

Comment (abubakar-sazim) [reply backend/nest-app/temp/User.js.json:1]: not sure bhai

File: backend/nest-app/src/app.controller.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+@Controller()
+export class AppController {
+  constructor(private authService: AuthService) {}
+
+  @UseGuards(JwtAuthGuard)
+  @Get('profile')
+  async getProfile(@Request() req) {
+    return req.user;
+  }
```
Comment (jUddinSazim) [inline backend/nest-app/src/app.controller.ts:13]: can we move it to the user controller?

File: backend/nest-app/src/auth/auth.service.ts [hunk 10]
```diff
@@ -1,12 +1,29 @@
...
+    }
+    return null;
   }
 
-  signin() {
-    return { msg: 'Signing In' };
+  async login(user: any) {
+    const payload = { email: user.email, name: user.name, sub: user.id };
+    return {
+      access_token: this.jwtService.sign(payload),
```
Comment (jUddinSazim) [inline backend/nest-app/src/auth/auth.service.ts:26]:
```suggestion
      accessToken: this.jwtService.sign(payload),
```

File: backend/nest-app/src/auth/strategies/local.strategy.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  constructor(authService: AuthService) {
+    super({
+      usernameField: 'email',
+    });
+    this.authService = authService;
+  }
+
+  async validate(email: string, password: string): Promise<any> {
+    const user = await this.authService.validateUser(email, password);
+    console.log(user);
```
Comment (jUddinSazim) [inline backend/nest-app/src/auth/strategies/local.strategy.ts:19]:
```suggestion
```

File: frontend/app/app/context/AuthContext.tsx
Comment (jUddinSazim) [inline frontend/app/app/context/AuthContext.tsx:1]: why commented file?

**Review by jUddinSazim — COMMENTED** [review]
I think you are troubling with using context in the FE

**Comment by jUddinSazim** [comment]
Here is a simple example of using context for auth https://www.mikealche.com/software-development/how-to-implement-authentication-in-next-js-without-third-party-libraries

File: frontend/app/context/AuthContext.tsx
Comment (jUddinSazim) [inline frontend/app/context/AuthContext.tsx:1]: may be we should keep the context out of app dir?

Comment (abubakar-sazim) [reply frontend/app/context/AuthContext.tsx:1]: yes bhai I kept the context folder out of the app dir

**Pull request merged**